### PR TITLE
[AIRFLOW-441] Add Python3 support to hipchat_operator

### PR DESCRIPTION
**Pull request by d-lee**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- _https://issues.apache.org/jira/browse/AIRFLOW-411_

Testing Done:
- Operator tested with Python 3.4

### CODE REVIEW

**Comment by d-lee** [comment]
I don't think that Travis CI build is broken by this commit. It looks like build job # 3625.4 is failing for all recent PRs.

**Comment by jlowin** [comment]
@d-lee rather than have to split the code for Py2/Py3, maybe just change `default_options.iteritems()` to `default_options.items()`, which will work in both versions? In this case I don't think there's any real benefit to using the lazy iterator in Py2 anyway.

**Comment by d-lee** [comment]
I don't see a reason for using iterator too but I was not completely sure why the original author wanted it this way, hence the code split. I am totally fine with just replacing it with `default_options.items()` and `params.items()` . 

Also, I don't think that utf-8 encoding for dictionary keys and values in original code actually works (line # 130). The result of an encode() is a bytestring, and bytestrings cannot be used as dictionary keys nor can they be serialized by json.dumps() . I think it has to be removed

**Comment by jlowin** [comment]
@d-lee yeah, the double-edged sword of `contrib` is that it's easy for the community to contribute but it's not necessarily maintained or tested as well as `core`. My guess with regard to `iteritems` is that the author was used to the best-practices Python2 way, and it's fine to switch. You may be right about the utf-8 encoding, I don't use this Operator and don't have an easy way to test it.

**Comment by jgao54** [comment]
@d-lee  @jlowin sorry about the confusion, I used .iteritems() out of habits, but definitely agree .items() is the way to go.

As for the question about why encoding was used,  'bytes' type was only introduced in Python3, so .encode() in python2 encodes 'unicode' to 'str'. 

In particular, this code is probably encoding 'str' to 'str' for the most part, and it works for Python2 for a very evil reason. It sees that the object is not a unicode and first try to implicitly coerce it into unicode (which works since it's ascii by default) and then encode it into string... I think the reason I did this was because I was hitting unicode errors during testing for some reason with data that were passed in, but I cannot recall exactly.

I do agree that for this to be python3 compatible we should not be encoding it. As long as removing the encode will still work with 2, Im more than happy for the change! Thanks so much for the fix.

**Comment by jgao54** [comment]
Just realized if 'notify' and 'card' were ever passed in to replace the default values, current implementation would break as well since they are not type 'str'. Oversight on my end. Thanks for the catch again.

**Comment by jlowin** [comment]
thanks @jgao54! No need to apologize, `iteritems` is the correct Py2 way :) 

Elsewhere in Airflow I used the [python-future](http://python-future.org/index.html) library for cross-version compatibility, in particular around strings. It's been a while since I looked at it but I think that if you do `from builtins import str` you should get a cross-version `str()` function that you can apply in place of that `.encode()` call. But please test it, that's just from memory.

**Comment by d-lee** [comment]
hipchat_operator.py is not tracked by codecov. Code quality decrease comes from changes brought in by rebasing on current master.

**Comment by jlowin** [comment]
@d-lee this looks good -- I assume you've tested in Py3? Is it possible to confirm that it runs in Py2 (I don't see why it wouldn't)?

**Comment by d-lee** [comment]
It's been tested in Python 3.4 (full integration test with operator used in a dag). Also we run everything on Py3, there is no quick and easy way for me to test in Py2. Sorry

**Comment by jlowin** [comment]
@jgao54 would it be possible for you to run this in Py2?

**Comment by bolkedebruin** [comment]
Why just not add some tests with mock? It will help verify py2 compat.

File: airflow/contrib/operators/hipchat_operator.py [hunk 1]
```diff
@@ -127,4 +126,4 @@ def prepare_request(self):
         self.method = 'POST'
         self.url = '%s/room/%s/notification' % (self.base_url, self.room_id)
         self.body = json.dumps(dict(
-            (k.encode('utf-8'), v.encode('utf-8')) for k, v in params.iteritems() if v))
+            (str(k), str(v)) for k, v in params.items() if v))
```
Comment (gwax) [inline airflow/contrib/operators/hipchat_operator.py:129]: dictionary comprehension?

``` python
{str(k): str(v) for k, v in params.items() if v}
```

File: airflow/contrib/operators/hipchat_operator.py [hunk 2]
```diff
@@ -110,7 +109,7 @@ def __init__(self, room_id, message, *args, **kwargs):
             'notify': False,
             'card': None
         }
-        for (prop, default) in default_options.iteritems():
+        for (prop, default) in default_options.items():
```
Comment (gwax) [inline airflow/contrib/operators/hipchat_operator.py:112]: For maximum compatibility, you might consider using `six`:

``` python
import six
...
for prop, default in six.iteritems(default_options)
```

**Comment by jlowin** [comment]
Such a minor change, LGTM

**Comment by jlowin** [comment]
Please note -- the correct issue is 411, not 441

File: airflow/contrib/operators/hipchat_operator.py (hunk 1 above)
Comment (gwax) [reply airflow/contrib/operators/hipchat_operator.py:129]: probably also a good idea to use `six.iteritems(params)`

**Pull request closed without merging**